0003-for-each.md

```diff
@@ -0,0 +1,634 @@
+}
+```
+
+A more serious shortcoming is that there is currently no efficient way to iterate over the elements of a `set` at all.
+The best alternative is to use the assign-such-that operator and set subtraction, illustrated in the reference manual as follows
+([sec 10.5.2](https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#1052-sets)):
+
+```dafny
+// s is a set<int>
+var ss := s;
+while ss != {}
+  decreases |ss|
+{
+  var i: int :| i in ss;
+  ss := ss - {i};
+  print i, "\n";
+}
+```
+
+This is functionally correct, but because set subtraction requires creating a fresh copy,
+when compiled this loop requires quadratic time and memory (or at least garbage-collection work) in the size of the set.
+The only way to avoid this cost is to write additional external code in the target language to interface directly with the
+internal runtime implementation of `set` values.
+
+By comparison, the equivalent `foreach` loop is:
+
+```dafny
+foreach i | i in s {
+  print i, "\n";
+}
+```
+
+The runtime implementation of this loop can use the native iteration features of the target language.
+
+Going further, it is better to avoid writing any imperative loop at all where possible. Sticking to immutable data and
+expressions rather than statements allows logic to be used in specifications as well as implementation,
+and reduces the burden on the Dafny verifier. The proposed sequence comprehension expression allows more
+logic that ultimately produces a sequence of values to be expressed as a value rather than a statement.
+Just producing a sequence of the values in a set is simple using a sequence comprehension expression: `seq i | i in s`.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## foreach Loops
+
+A `foreach` loop in Dafny resembles `for` or `foreach` loops in many other mainstream languages,
+but is substantially more general and expressive. A very simple example only looks slightly different than expected:
+
+```dafny
+foreach x | x in [1, 2, 3, 4, 5] {
+  print x, "\n";
+}
+```
+
+The feature supports much more sophisticated uses as well, however,
+including binding multiple variables at once and filtering:
+
+```dafny
+var myDoubleDeckerSeq: seq<seq<int>> := ...;
+foreach x, y | x in myDoubleDeckerSeq && y in x && y != 0 {
+  Process(y);
+}
+```
+
+The high-level syntax for a `foreach` loop reuses several existing concepts:
+
+```
+ForeachLoop ::=
+  "foreach"
+  QuantifierDomain
+  LoopSpec
+  Body
+```
+
+More specifically, this expands to:
+
+```
+ForeachLoop ::=
+  "foreach" 
+  Ident [":" Type] { "," Ident [":" Type] } [ "|" Expression ]
+  { InvariantDecl | ModifiesDecl | DecreasesDecl }
+  Body
+```
+
+A `foreach` loop closely resembles a `forall` statement, the key difference being that a `forall`
+loop executes its body once for every tuple of quantified values simultaneously, whereas a `foreach` loop
+executes its body once for each tuple of quantified values in sequence, one at a time.
```

Comment:
Note that `forall` only consist of these restrictions:
* A forall loop is always ghost (someone can confirm that? I have never seen a forall statement in a compiled context)
* A forall statement cannot contains statements that mutate variables.
* A forall statement cannot have breaks or continue.

```dafny
  ghost var result: seq<int> := []; // Would not 
  forall x | x in [1, 2, 3, 4, 5] {
   result := result + [x]; // Error
   continue; // Error
  }
```

Hence, with your proposal, we could simply make "forall" obsolete because foreach iterations that are in a ghost context, don't mutate variables, and without break or continue would have the SAME semantics as a forall statement.
Maybe the only difference would be the infiniteness. Forall would work for an infinite collection, but foreach wouldn't, right?

Reply:
No you can use `forall` in compiled contexts, but only in very limited circumstances. https://dafny-lang.github.io/dafny/DafnyRef/DafnyRef#sec-forall-statement describes how there are three distinct cases of `forall` statements supported, and only the first can be used outside of ghost code.

I'm proposing `foreach` only allows a potentially infinite domain if not used in ghost code, and with either `decreases *` or an alternate termination measure. You make an interesting point though: the semantics of `forall` are that the body is executed simultaneously for every tuple, but if there is no interaction between these executions, that's indistinguishable from executing them sequentially.

For now I at least lean towards adding `foreach` while keeping `forall`, rather than allowing `forall` to be used in more cases: if nothing else I think `foreach` is a much better name for a loop where iterations CAN observe the effects of previous iterations. But I'll put some more thought into this.

Reply:
Note that the `forall` statement is _not_ a loop. It is an _aggregate statement_. Its compiled form is like a generalization of simultaneous assignment. The beauty of the `forall` statement is that you use it _without a loop invariant_. :O This is made possible because--again--the `forall` statement is not a loop in the first place!

Loops (`while`, `for`, `foreach`) sequence their iterations. And to use them, you need a loop invariant, so that you can capture "what has been done so far".